dm/dm/master/shardddl/optimist.go

```diff
@@ -593,9 +593,30 @@ func (o *Optimist) handleSourceTables(ctx context.Context, sourceCh <-chan optim
 			if !ok {
 				return
 			}
-			updated := o.tk.Update(st)
-			o.logger.Info("receive source tables", zap.Stringer("source tables", st),
-				zap.Bool("is deleted", st.IsDeleted), zap.Bool("updated", updated))
+			o.mu.Lock()
+			addedTable, droppedTable := o.tk.Update(st)
+			// handle create table
+			for routeTable := range addedTable {
+				lock := o.lk.FindLock(utils.GenDDLLockID(st.Task, routeTable.DownSchema, routeTable.DownTable))
+				if lock != nil {
+					lock.AddTable(st.Source, routeTable.UpSchema, routeTable.UpTable, true)
+				}
+			}
+			// handle drop table
+			for routeTable := range droppedTable {
+				lock := o.lk.FindLock(utils.GenDDLLockID(st.Task, routeTable.DownSchema, routeTable.DownTable))
+				if lock != nil {
+					cols := lock.TryRemoveTable(st.Source, routeTable.UpSchema, routeTable.UpTable)
+					if !lock.HasTables() {
+						o.lk.RemoveLock(lock.ID)
+					}
+					_, err := optimism.DeleteInfosOperationsTablesByTable(o.cli, st.Task, st.Source, routeTable.UpSchema, routeTable.UpTable, lock.ID, cols)
```

Comment:
Not important: do we need to batch these delete operations and run them out of `o.mu.Lock` part? I'm afraid this will cause this lock is held for a long time.

Reply:
I think it's ok because this line usually run when drop a table. For stop-task, the meta data in etcd already delete at https://github.com/pingcap/tiflow/blob/21dedc45c4022be849b575ef12b4fdcd089963ba/dm/dm/master/server.go#L642, hope it will not cause much time. I will add a todo.